Add support Redis lock system for commands

kuzmany opened the pull request:
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | "features" for all features, enhancements and bug fixes (until 3.3.0 is released) <!-- see below -->
| Bug fix?                               | yes/no
| New feature?                           | yes/no
| Deprecations?                          | yes/no
| BC breaks?                             | yes/no
| Automated tests included?              | yes/no
| Related user documentation PR URL      | mautic/mautic-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--f
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#step-5-work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "features" branch.
-->

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do.
-->
#### Description:
Mautic4 for bring support for Lock component. Now it's easy to add any of supported lock systems to our commands. This PR bring support for redis 

<!--
If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Setup Redis. On ddev you can zse https://github.com/mautic/mautic/pull/10347
3. Setup it to local.php 

```php
      'cache_adapter_redis'     => [
            'dsn'     => 'redis://redis',
        ],

```
4. Run `php bin/console m:s:r -x redis`
5. Should work like before

<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->


<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10347"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

kuzmany committed:
Add support for Redis lock system

kuzmany committed:
Add unit tests

kuzmany committed:
Fix typo dsn parameter

kuzmany committed:
Fix unit tests

kuzmany committed:
Remove not used variable

mollux commented on app/bundles/CoreBundle/Command/ModeratedCommand.php:
```diff
@@ -180,7 +183,18 @@ private function checkPid(): bool
+            case self::MODE_REDIS:
+                $dsn   = $this->getContainer()->get('mautic.helper.core_parameters')->get('cache_adapter_redis')['dsn'] ?? null;
+                $redis = new \Redis();
+                $redis->connect(parse_url($dsn, PHP_URL_HOST));
```
@kuzmany the DSN can contain a socket path, so the `parse_url` won't work in that case

kuzmany replied:
Good point 🥇 I have to check it again

kuzmany replied:
I rewrite it with support RedisAdapter , options and lock timeout

kuzmany committed:
Add handle redis with options and lock time

kuzmany committed:
Fix unit tests

kuzmany committed:
Remove check for Redis class from tests

mabumusa1 commented:
@kuzmany we need to bring this to M5 please

kuzmany committed:
Fix CoreParametersHelper

kuzmany commented:
@mabumusa1 I rebased it to 5.x but we added a new service CoreParametersHelper to ModerateCommand, then need the change apply to all commands and fix all unit tests related to the command.  Need more time for it.

kuzmany committed:
Add CoreParametersHelper to each ModerateCommand

mabumusa1 commented:
@kuzmany there are two failing tests, please fix them so I can test.

kuzmany committed:
Fix unused variable CoreParametersHelper

kuzmany committed:
Fix unit tests

kuzmany committed:
Fix unit test

mabumusa1 reviewed (approved):
Code change looks good, I tested locally and lock worked as expected

volha-pivavarchyk reviewed (approved):
The command works as before

luizeof commented:
Hello friends @RCheesley @escopecz , how are you?

Does this lock change modify the moderationKey ?

Mautic doesn't allow us to run a command with the --id parameter at the same time, which is great.

However, if we pass --min-contact-id or --max-contact-id, separating the contacts by groups to speed up the sending in parallel, the lock is also activated.

is there any chance to add --min-contact-id and --max-contact-id to moderationKey ?

Thank you so much for releasing 5 alpha with all your efforts to make this tool even more amazing.

luizeof commented:
PS: https://github.com/mautic/mautic/pull/11549 solves a lot of things and already guarantees a lot of agility. it's more about the range of contacts.

escopecz commented:
@luizeof great question! We'll be documenting this soon. Mautic 5 will send emails via [Symfony Messenger](https://symfony.com/doc/5.4/messenger.html) and so you'll be able to run as many workers to send emails concurrently as you wish. No need to split contacts into groups by ID ranges. I can't wait to read about how the community will manage auto-scaling of such workers to bring the server cost down and send emails by warp speed 🙂 🚀
